src/utils/daos_hdlr.c

```diff
@@ -2646,45 +2648,421 @@ set_dm_args_default(struct dm_args *dm)
+	buffers = calloc(num_attrs, sizeof(void*));
+	if (buffers == NULL) {
+		fprintf(stderr, "failed to allocate user attribute buffer");
+		rc = 1;
+		D_GOTO(out, rc);
+	}
+
+	/* Create the array of names */
+	cur_size = 0;
+	for (i = 0; i < num_attrs; i++) {
+		name_len = strnlen(name_buf + cur_size, total_size - cur_size);
+		if (name_len == total_size - cur_size) {
+			/* end of buf reached but no end of string, ignoring */
+			break;
+		}
+		names[i] = strndup(name_buf + cur_size, name_len + 1);
+		cur_size += name_len + 1;
+	}
+
+	/* Get the buffer sizes */
+	rc = daos_cont_get_attr(coh, num_attrs, (const char* const*)names,
+				NULL, sizes, NULL);
+	if (rc != 0) {
+		fprintf(stderr, "failed to get user attribute sizes "DF_RC,
+			DP_RC(rc));
+		rc = 1;
+		D_GOTO(out, rc);
+	}
+
+	/* Allocate space for each value */
+	for (i = 0; i < num_attrs; i++) {
+		buffers[i] = calloc(sizes[i], sizeof(size_t));
+		if (buffers[i] == NULL) {
+			fprintf(stderr, "failed to allocate user attribute "
+				"buffer");
+			rc = 1;
+			D_GOTO(out, rc);
+		}
+	}
+
+	/* Get the attribute values */
+	rc = daos_cont_get_attr(coh, num_attrs, (const char* const*)names,
+				(void * const*)buffers, sizes, NULL);
+	if (rc != 0) {
+		fprintf(stderr, "failed to get user attribute values "DF_RC,
+			DP_RC(rc));
+		rc = 1;
+		D_GOTO(out, rc);
+	}
+
+	/* Return values to the caller */
+	*_n = num_attrs;
+	*_names = names;
+	*_buffers = buffers;
+	*_sizes = sizes;
+out:
+	if (rc != 0) {
+		dm_cont_free_usr_attrs(num_attrs, &names, &buffers, &sizes);
+	}
+	D_FREE(name_buf);
+	return rc;
+}
+
+/* Copy all user attributes from one container to another.
+ * Returns 1 on error, 0 on success. */
+int
+dm_copy_usr_attrs(daos_handle_t src_coh, daos_handle_t dst_coh)
+{
+	int	num_attrs = 0;
+	char	**names = NULL;
+	void	**buffers = NULL;
+	size_t	*sizes = NULL;
+	int	rc;
+
+	/* Get all user attributes */
+	rc = dm_cont_get_usr_attrs(src_coh, &num_attrs,
+				   &names, &buffers, &sizes);
+	if (rc != 0) {
+		rc = 1;
+		D_GOTO(out, rc);
+	}
+
+	if (num_attrs == 0) {
+		rc = 0;
+		D_GOTO(out, rc);
+	}
+	rc = daos_cont_set_attr(dst_coh, num_attrs, (char const* const*) names,
+				(void const* const*) buffers, sizes, NULL);
+
+	if (rc != 0) {
+		fprintf(stderr, "Failed to set user attrs: "DF_RC, DP_RC(rc));
+		rc = 1;
+		D_GOTO(out, rc);
+	}
+
+out:
+	dm_cont_free_usr_attrs(num_attrs, &names, &buffers, &sizes);
+	return rc;
+
+}
+
+/*
+ * Get the container properties for a container in a format similar
+ * to what daos_cont_set_prop expects.
+ * The last entry is the ACL and is conditionally set only if
+ * the user has permissions.
+ * The properties need to remain in the order expected by serialization.
+ * Returns 1 on error, 0 on success.
+ */
+int
+dm_cont_get_all_props(daos_handle_t coh, daos_prop_t **_props, bool get_oid)
+{
+	int		rc;
+	daos_prop_t	*props = NULL;
+	daos_prop_t	*prop_acl = NULL;
+	daos_prop_t	*props_merged = NULL;
+	uint32_t	num_props = 15;
+
+    	if (get_oid) {
```

Comment:
(style)  code indent should use tabs where possible
(style)  please, no space before tabs